### ISSUE TITLE
SampleEval Optimizer

### ISSUE THREAD

**Issue by okennedy** [UBOdin/mimir#216]
Background: 
* [MCDB](http://www.academia.edu/download/46769745/mcdb.pdf)
* [TupleBundler](https://github.com/UBOdin/mimir/blob/master/src/main/scala/mimir/exec/TupleBundler.scala)

At present, Mimir employs two query evaluation strategies: Best-Guess and Sample-Based.  Best-Guess plugs in a "most likely" candidate value for query evaluation, while Sample-Based samples a set of [Possible Worlds](https://github.com/UBOdin/mimir/wiki/Concepts-CTables#possible-worlds-semantics) and evaluates the query in parallel on each.  This project involves Sample-Based evaluation.

----

Sample-Based Evaluation can be further subdivided into two approaches: Wide and Tall.  Consider for example the table (where x is a random variable):
```
R | A  B
--+------
  | 1 {x}
```
If we wanted to take 3 samples (with `{x}` evaluating to 5, 6, and 7 respectively), we could represent the samples through added columns
```
R | A  B1  B2  B3
--+-------------
  | 1  5  6  7
```
This approach means we'd also need a way to represent which samples were "active", to allow for filtering.  For example, with `Q = SELECT[B <> 6]`:
```
Q | A  B1  B2  B3  WORLDS
--+-------------
  | 1  5  6  7  00...00101
```
Here, `WORLDS` is a bitvector with each bit corresponding to one possible world.  SELECT unsets bits (e.g., the world B2 where `B = 6`).  We call this the **wide** representation

Alternatively, we could represent the samples through added rows, annotating each row with a "World ID"
```
R | A B WORLD_ID
--+----------
  | 1 5   1
  | 1 6   2
  | 1 7   3
```
We call this the **tall** representation.

------

Each representation has benefits and drawbacks: 
* The wide representation tends to be more compact, particularly if there are relatively few uncertain columns, and should be easier for most engines to pipeline, since we don't need UNIONs to construct them.
* The wide representation is more efficient in general, as we don't need a post-processing aggregation step to gather together instances of the same output row.
* The tall representation is required for aggregate queries when a group-by attribute is uncertain.
* The tall representation should be more friendly for equi-join queries when a join attribute is uncertain, as the wide representation creates gnarly join conditions that make it difficult for an optimizer to create a Hash join.

------

In short, for any given query, we may want to have a different set of representations (and possibly multiple representations in a single query).  For this project, we'd like to see an experimental evaluation of the tradeoffs (when is Wide better, and when is Tall better?), and an optimizer built into Mimir that selects an optimal sampling-based evaluation plan for a given query.
